### PR TITLE
add overload `add(a: var string, b: openArray[char])`

### PR DESCRIPTION
**Pull request by timotheecour**
2 goals:
* it's currently hard to convert openArray[char] to a string (unless you do it manually, essentially inlining the code in this PR)
* avoid allocations when possible if user already has a `y` openArray to begin with

## future work
- [ ] after merged, fix this comment:
https://github.com/nim-lang/Nim/pull/15421/files#r522703485
- [ ] see also `addCstringN`
- [x] see https://github.com/nim-lang/Nim/pull/15951#issuecomment-785329824 (see also https://github.com/nim-lang/Nim/pull/17478)

### CODE REVIEW

File: changelog.md [hunk 1]
```diff
@@ -35,6 +35,8 @@
 
 - `doAssertRaises` now correctly handles foreign exceptions.
 
+- Added overload `ssytem.add(a: var string, b: openArray[char])`
```
Comment (juancarlospaco) [inline changelog.md]:
```suggestion
- Added overload `system.add(a: var string, b: openArray[char])`.
```

Comment (timotheecour) [reply changelog.md]: thx, fixed

File: changelog.md [hunk 2]
```diff
@@ -35,7 +35,7 @@
 
 - `doAssertRaises` now correctly handles foreign exceptions.
 
-- Added overload `ssytem.add(a: var string, b: openArray[char])`
+- Added overload `sytem.add(a: var string, b: openArray[char])`
```
Comment (juancarlospaco) [inline changelog.md]:
```suggestion
- Added overload `system.add(a: var string, b: openArray[char])`.
```

Comment (timotheecour) [reply changelog.md]: yikes... fixed

File: lib/system.nim [hunk 3]
```diff
@@ -1032,6 +1032,18 @@ proc add*(x: var string, y: string) {.magic: "AppendStrStr", noSideEffect.}
...
   ##   assert(tmp == "abcd")
 
+proc add*(x: var string, y: openArray[char]) =
+  ## Concatenates `x` and `y` in place. `y` must not overlap with `x` to
+  ## allow future memcpy optimizations.
+  let n = x.len
+  x.setLen n + y.len
+    # pending https://github.com/nim-lang/Nim/issues/14655#issuecomment-643671397
+    # use x.setLen(n + y.len, isInit = false)
+  var i=0
```
Comment (narimiran) [inline lib/system.nim]:
```suggestion
  var i = 0
```

File: lib/system.nim [hunk 4]
```diff
@@ -1032,6 +1032,18 @@ proc add*(x: var string, y: string) {.magic: "AppendStrStr", noSideEffect.}
...
+  while i<y.len:
```
Comment (narimiran) [inline lib/system.nim]:
```suggestion
  while i < y.len:
```

**Comment by metagn** [comment]
While the intention here is good there are so many things in the stdlib that could benefit from using a more generic char sequence type (like in strutils) and openarray[char] is not the best candidate, lest we snowball into doing this everywhere. This change is however small and mirrors add(var seq[T], openarray[T]), so it's fine, except that that overload doesn't copy memory directly and uses move/sink instead

**Comment by juancarlospaco** [comment]
A snowball has to start somewhere... :smirk_cat:

**Comment by timotheecour** [comment]
> openarray[char] is not the best candidate

string/seq/array implicitly converts to openArray[char], which is the main point; I'm not sure what your other "best candidate" is (apart from std/views see https://github.com/nim-lang/Nim/pull/14869 which has the advantage of being 1st class but doesn't implicitly convert from string/seq/array; only explicitly)

> mirrors add(var seq[T], openarray[T]), so it's fine, except that that overload doesn't copy memory directly and uses move/sink instead

for openArray[char] (this PR) memcpy (or memmove if memory overlaps) is the correct thing to do; for the other existing overload with seq[T], we could/should do the same depending on `T`

**Review by ringabout — APPROVED** [review]

**Comment by Araq** [comment]
Can we add this to the new `strbasics` instead?

**Comment by timotheecour** [comment]
> Can we add this to the new strbasics instead?

done
@Araq PTAL

I'll followup with a PR to address this:
> xxx use `nimCopyMem(x[n].addr, y[0].addr, y.len)` after some refactoring

with some refactoring from compiler/strutils2

**Pull request merged**